src/symbol/quads.js

```diff
@@ -228,11 +242,30 @@ export function getGlyphQuads(anchor: Anchor,
                        feature: Feature,
                        imageMap: {[_: string]: StyleImage},
                        allowVerticalPlacement: boolean): Array<SymbolQuad> {
+    const quads = [];
+    if (shaping.positionedLines.length === 0) return quads;
 
     const textRotate = layer.layout.get('text-rotate').evaluate(feature, {}) * Math.PI / 180;
-    const quads = [];
+    const rotateOffset = getRotateOffset(textOffset);
 
+    let shapingHeight = Math.abs(shaping.top - shaping.bottom);
     for (const line of shaping.positionedLines) {
+        shapingHeight -= line.lineOffset;
+    }
+    const lineCounts = shaping.positionedLines.length;
+    const lineHeight = shapingHeight / lineCounts;
+    const getMidlineOffset = function(shaping, lineHeight, previousOffset, lineIndex) {
+        const currentLineHeight = (lineHeight + shaping.positionedLines[lineIndex].lineOffset);
+        if (lineIndex === 0) {
+            return previousOffset + currentLineHeight / 2.0;
+        }
+        const aboveLineHeight = (lineHeight + shaping.positionedLines[lineIndex - 1].lineOffset);
+        return previousOffset + (currentLineHeight + aboveLineHeight) / 2.0;
+    };
```

Comment:
Nit: functions that don't need to be a closure (nested within another function) should be moved to the top scope.